Attr assign backward compatible rebase

yuewko opened the pull request:
This PR include all the changes to add numerical functions in support of auto-scaling policies.

It also include changes that allow Attribute Assignment Expression to support both Immediate Values as well as Expressions

vasili-v commented on Makefile:
```diff
@@ -154,6 +154,10 @@ build-egen: build-dir
 test-pdp: cover-out
 	$(AT)/pdp && $(GOTESTRACE)
 
+.PHONY: test-pdp-integration
```
I don't see a need of the integration tests it the form they have been added here. Themis package includes several different things - pdp, parsers, server, client libraries. What do the tests integrate? Why only some aspects of numbers, but not whole thing? I see those more like parser tests so please move them to related parser directory.

yuewko replied:
Maybe **integration** is a poor choice of word, but these are definitely not **unit** tests either. They don't just test the parsing, but the actual calculation too.

On your other point, I agree they definitely don't test *all* aspects of numbers. Please let me know what are the significant omissions and I'll definitely add those too.

vasili-v replied:
I prefer to test calculations inside pdp constructing policies with Make* functions. Basically, we don't need much integration inside single repository.

yuewko replied:
OK. Let me give it a shot.

vasili-v commented on pdp/ast/jast/obligations.go:
```diff
@@ -3,14 +3,15 @@ package jast
 import (
 	"encoding/json"
 
+	//	"fmt"
```
Don't comment it - just remove

yuewko replied:
My bad. Will fix.

yuewko replied:
Done

vasili-v commented on pdp/ast/jast/obligations.go:
```diff
@@ -24,11 +25,43 @@ func (ctx context) unmarshalObligationItem(d *json.Decoder) (pdp.AttributeAssign
+			if err != nil {
+				return err
+			}
+			/*
```
No need to comment code - just remove it. If you need it for later make other branch.

yuewko replied:
My bad. It's no longer needed. I'll remove it.

yuewko replied:
Done

vasili-v commented on pdp/ast/jast/obligations.go:
```diff
@@ -24,11 +25,43 @@ func (ctx context) unmarshalObligationItem(d *json.Decoder) (pdp.AttributeAssign
 		}
 
-		e, err = ctx.unmarshalValueByType(a.GetType(), d)
+		delim, val, err := jparser.CheckObjectStartOrValue(d, "argument")
```
As I recall, we've agreed on no backward compatibility. So obligation is always an expression. It makes no need of CheckObjectStartOrValue function.

yuewko replied:
Yes, I know... I was very close to making it work when we discussed, so I dug deep. I thought it would be a good goal to avoid PAP changes.

My original plan was to have the numerical function changes (using expression only of obligations) as one PR, and the change to make it backward compatible a separate PR. This is so you can review to see whether the changes for backward compatibility is acceptable. (unfortunately, there was a lot conflicts when rebasing, so I bailed out...)

If you feel strongly that the backward compatibility changes should be pulled, I'll do that.

vasili-v commented on pdp/ast/jast/obligations.go:
```diff
@@ -24,11 +25,43 @@ func (ctx context) unmarshalObligationItem(d *json.Decoder) (pdp.AttributeAssign
+				return bindError(err, k)
+			}
+		} else if delim == jparser.DelimArrayStart {
+			val, err := jparser.GetArray(d, "obligations")
```
JAST is a stream parser. You need to work on token level. Don't try to get whole array and then parse it. Instead parse items as they appear with `CheckArrayStart` and by token iterations. However, if we follow no backward compatibility this isn't needed. Because obligation is always expression and you can just use existing expression parser here.

yuewko replied:
Yes, totally understood. (please also see https://github.com/infobloxopen/themis/pull/112#discussion_r164430421)

vasili-v commented on pdp/ast/yast/obligations.go:
```diff
@@ -23,9 +23,20 @@ func (ctx context) unmarshalObligationItem(v interface{}) (pdp.AttributeAssignme
+	var e pdp.Expression
+	m, err = ctx.validateMap(v, "obligation assignment")
+	if err == nil {
+		// if it's a map, assume it's an expression
```
Let's stick to no backward compatibility. It allows in addition to other simplifications to bypass some corner cases related to the assumption.

yuewko replied:
Sounds good - I'll take those code out.

BTW, I just remembered - one reason I tried to make it backward compatible is because the XACML specification stipulates using immediate value. I'm good if you don't feel that's critical.

Thanks!!

vasili-v commented on pdp/attribute.go:
```diff
@@ -23,6 +23,10 @@ const (
 	TypeBoolean
 	// TypeString is string data type.
 	TypeString
+	// TypeInteger is integer data type.
```
Please add tests to attribute_test.go so we keep whole type converters tested.

yuewko replied:
Done

vasili-v commented on pdp/attribute.go:
```diff
@@ -403,6 +463,12 @@ func (v AttributeValue) Serialize() (string, error) {
+		return strconv.FormatInt(v.v.(int64), 10), nil
+
+	case TypeFloat:
+		return strconv.FormatFloat(v.v.(float64), byte('G'), 16, 64), nil
```
If no specific objections the -1 precision is better as it makes `FormatFloat` so many digits that `ParseFloat` returns exactly value `FormatFloat` gets. And you don't need to cast 'G' to byte. The changes make the line to look exactly as [above](https://github.com/infobloxopen/themis/pull/112/files#diff-f4fc264e7d4c0af26700a6f5fc900b41R278).

yuewko replied:
Yes - my bad, again. I thought I changed them all to `-1`.
Will fix.

yuewko replied:
Done

vasili-v commented on pdp/context.go:
```diff
@@ -192,6 +202,41 @@ func (c *Context) calculateStringExpression(e Expression) (string, error) {
+	return v.float()
+}
+
+func (c *Context) calculateFloatOrIntegerExpression(e Expression) (float64, error) {
```
I'd like to see explicit conversion for each type. This assumes int-to-float implicit conversion. Having expression by argument type function selection on parsing stage such implicit conversion makes hard to follow function resolution rules. And we need to document it clearly as C++ does. I'd like to not mess with it.

yuewko replied:
I understand your concern.

I just wanted to point out that `calculateFloatOrIntegerExpression` is specifically designed for the float numerical functions, where I specifically want to promote int arguments to float. Without it, the code will be duplicated for each of the arguments for each of the float functions.

Maybe I can move/rename that function if you feel it doesn't below to that group of `calculate` functions.

Please let me know how you'd like to proceed.

vasili-v commented on pdp/target.go:
```diff
@@ -192,7 +192,16 @@ type twoArgumentsFunctionType func(first, second Expression) Expression
+			TypeString: makeFunctionStringEqual},
+		TypeInteger: {
+			TypeInteger: makeFunctionIntegerEqual},
+		TypeFloat: {
```
Actually, here you missed float+int and int+float -> makeFunctionFloatEqual cases if you want to follow implicit conversion. However, I don't think it's worth to make it.

yuewko replied:
Sounds good!

Actually, I thought about that. The options are:
- Promote the int to float and then compare as floats. This typically is not reliable. It might makes sense for greater than, but not for equal, IMHO.
- Cast/Round the float to int and then compare as int. This is potentially very confusing for the users.

yuewko commented:
https://github.com/infobloxopen/themis/pull/112/commits/02aa45c562e095884f4196f81159567ed72e149d addressed the following review comments:

- https://github.com/infobloxopen/themis/pull/112#pullrequestreview-92190947
- https://github.com/infobloxopen/themis/pull/112#pullrequestreview-92191337
- https://github.com/infobloxopen/themis/pull/112#pullrequestreview-92203254
- https://github.com/infobloxopen/themis/pull/112#pullrequestreview-92207587

yuewko commented:
Added unit tests for numerical functions to address https://github.com/infobloxopen/themis/pull/112#pullrequestreview-92190461

yuewko commented:
@vasili-v , I believe I have addressed all the review comments, with the exception of documentation updates (i.e. README and Examples).

Please review when you have a chance.

yuewko commented:
@vasili-v , the latest commit includes README updates and additional examples. This should have addressed all your comments.

Thanks!!

yuewko commented:
OK. Thanks for looking at it.
